pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -229,6 +255,266 @@ func (p *Pod) Skip() bool {
 	return false
 }
 
+func (p *Pod) groupName() string {
+	return p.Object().GetLabels()[GroupNameLabel]
+}
+
+func (p *Pod) groupTotalCount() (int, error) {
+	if p.groupName() == "" {
+		return 0, fmt.Errorf("pod doesn't have a '%s' label", GroupNameLabel)
+	}
+
+	gtcAnnotation, ok := p.Object().GetAnnotations()[GroupTotalCountAnnotation]
+	if !ok {
+		return 0, fmt.Errorf("failed to extract '%s' annotation",
+			GroupTotalCountAnnotation)
+	}
+
+	gtc, err := strconv.Atoi(gtcAnnotation)
+	if err != nil {
+		return 0, err
+	}
+
+	if gtc < 1 {
+		return 0, fmt.Errorf("incorrect annotation value '%s=%s': group total count should be greater than zero",
+			GroupTotalCountAnnotation, gtcAnnotation)
+	}
+
+	return gtc, nil
+}
+
+func (p *Pod) constructGroupPodSets(podsInGroup corev1.PodList) ([]kueue.PodSet, error) {
+	groupTotalCount, err := p.groupTotalCount()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		resultPodSets []kueue.PodSet
+	)
+
+	for _, podInGroup := range podsInGroup.Items {
```

Comment:
Since there is no guarantee about the order of the pods, should we sort the podsets by hash at the end? This might guarantee that the same set of pods are not considered unmatching with the Workload object.